src/d3a/models/market/one_sided.py

```diff
@@ -90,92 +94,115 @@ def delete_offer(self, offer_or_id: Union[str, Offer]):
-                        source_rate = trade_rate / (1 + self.transfer_fee_ratio) \
-                                      - self.transfer_fee_const
-                        self.market_fee += (trade_rate - source_rate) * energy
-                    else:
-                        source_rate = trade_rate
-                    accepted_offer = Offer(
-                        accepted_offer_id,
-                        source_rate * energy,
-                        energy,
-                        offer.seller,
-                        offer.market
-                    )
-                    residual_offer = Offer(
-                        str(uuid.uuid4()),
-                        offer.price / offer.energy * (offer.energy - energy),
-                        offer.energy - energy,
-                        offer.seller,
-                        offer.market
-                    )
-                    self.offers[residual_offer.id] = residual_offer
-                    log.info(f"[OFFER][CHANGED][{self.time_slot_str}] "
-                             f"{original_offer} -> {residual_offer}")
-                    offer = accepted_offer
-
-                    self.bc_interface.change_offer(offer, original_offer, residual_offer)
-                    self._sorted_offers = sorted(self.offers.values(),
-                                                 key=lambda o: o.price / o.energy)
-                    self._notify_listeners(
-                        MarketEvent.OFFER_CHANGED,
-                        existing_offer=original_offer,
-                        new_offer=residual_offer
-                    )
-                elif energy > offer.energy:
-                    raise InvalidTrade("Energy can't be greater than offered energy")
-                else:
-                    # Requested energy is equal to offer's energy - just proceed normally
-                    if trade_rate is not None:
-                        # offered_price adjusted to trade rate
-                        if iaa_fee:
-                            # adjusted in grid fee
-                            offer.price = trade_rate * offer.energy / \
-                                          (1 + self.transfer_fee_ratio) + self.transfer_fee_const
-                            self.market_fee += \
-                                trade_rate * (self.transfer_fee_ratio) * offer.energy
-                        else:
-                            offer.price = trade_rate * offer.energy
+
+            energy_portion = energy / offer.energy
+            if energy == 0:
+                raise InvalidTrade("Energy can not be zero.")
+            # partial energy is requested
+            elif energy < offer.energy:
+                original_offer = offer
+                accepted_offer_id = offer.id \
+                    if self.area is None or self.area.bc is None \
+                    else offer.real_id
+
+                assert trade_rate + FLOATING_POINT_TOLERANCE >= (offer.price / offer.energy)
```

Comment:
I dont understand this assert at this point of the code. Is there any reason for not putting it just after  L 137?

Reply:
You are correct, it is better to place the assignment and validation close. I will move these probably to a separate function too, in order to reduce a bit the size of the accept_offer method.